### PR TITLE
[ASR] Support audio_filepath as a list of single-channel files

### PR DESCRIPTION
**Pull request by anteju**
# What does this PR do ?

Currently, multichannel audio can be used in `audio_filepath` by providing a path to a multichannel file.
This PR adds support for `audio_path` to specify multichannel audio by providing a list of paths to single-channel files.

**Collection**: ASR

# Changelog 

- `AudioSegment.from_file` can load multichannel audio from a list of paths to single-channel audio files
- `manifest.get_full_path` supports lists as input and can resolve path to each path in a list
- Added unit tests for `AudioSegment`
	- `TestAudioSegment:: test_audio_segment_from_file` for existing functionality, loading a single file
	- `TestAudioSegment:: test_audio_segment_multichannel_with_list ` for new functionality, loading multiple single-channel files
- Added unit tests for `manifest.get_full_path`
	- `TestPreprocessingUtils:: test_get_full_path_local` when resolving local paths, both single paths and lists
	- `TestPreprocessingUtils:: test_get_full_path_ais` when resolving paths pointing to AIStore, both single paths and lists

# Usage

## Example scripts
- [examples.zip](https://github.com/NVIDIA/NeMo/files/10887310/examples.zip)

## Description

Currently, ASR models can be used to transcribe multichannel audio by providing a path to a multichannel audio file in `audio_filepath` and selecting a single channel using `channel_selector`.
This PR adds an option to set `audio_filepath` to a list of paths to individual channels, each containing a single channel. This is useful in situations where multichannel audio is provided as individual channels.

Two examples with `audio_filepath` containing a list of paths to single-channel files are provided below.
The first example is directly using `transcribe()` od an `ASRModel`. The second example is running `speech_to_text_eval.py`.

### Example 1

This example will check if `asr_model.transcribe` is working the same when a path to multichannel audio is provided as

- a path to a multichannel audio file (this is currently supported)
- a list of paths to single-channel audio files (added in this PR)

In both cases, channel_selector can be used to pick one channel from the input multichannel audio.

The test will run with two different models: `stt_en_conformer_ctc_small` and `stt_en_conformer_transducer_small` and each of the 3 available channels.

To run the test:

```
python test_asr_model_transcribe.py
```

### Example 2

This example will check if `speech_to_text_eval.py` is working the same when a path to multichannel audio is provided as

- a path to a multichannel audio file (this is currently supported)
- a list of paths to single-channel audio files (added in this PR)

In both cases, channel_selector can be used to pick one channel from the input multichannel audio.

The test will run with two different models: `stt_en_conformer_ctc_small` and `stt_en_conformer_transducer_small` and each of the 3 available channels.

To run the test:

```
python test_speech_to_text_eval.py --nemo-dir PATH_TO_NEMO_DIR
```

# Before your PR is "Ready for review"
**Pre checks**:
- [x] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [x] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [x] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: examples/asr/transcribe_speech.py [hunk 1]
```diff
@@ -105,7 +105,9 @@ class TranscriptionConfig:
     pretrained_name: Optional[str] = None  # Name of a pretrained model
     audio_dir: Optional[str] = None  # Path to a directory which contains audio files
     dataset_manifest: Optional[str] = None  # Path to dataset's JSON manifest
-    channel_selector: Optional[int] = None  # Used to select a single channel from multi-channel files
+    channel_selector: Optional[
+        Union[int, str]
```
Comment (titu1994) [inline examples/asr/transcribe_speech.py:109]: Hmm dunno about this, hydra doesn't support optional union and we don't test this script in CI (or maybe we do with an old QN model). Did you test this locally ?

File: nemo/collections/asr/parts/preprocessing/segment.py [hunk 2]
```diff
@@ -199,51 +201,111 @@ def from_file(
...
-                        samples = f.read(int(duration * sample_rate), dtype=dtype)
-                    else:
-                        samples = f.read(dtype=dtype)
-            except RuntimeError as e:
-                logging.error(
-                    f"Loading {audio_file} via SoundFile raised RuntimeError: `{e}`. "
-                    f"NeMo will fallback to loading via pydub."
+        if isinstance(audio_file, list):
+            # If a list of files is provided, load samples from individual single-channel
+            # files and concatenate them along the channel dimension
```
Comment (titu1994) [inline nemo/collections/asr/parts/preprocessing/segment.py:209]: What if user names mistakes and send list of two different file here ? Can we detect and raise error?

File: nemo/collections/asr/parts/utils/transcribe_utils.py [hunk 3]
```diff
@@ -227,10 +226,8 @@ def prepare_audio_data(cfg: DictConfig) -> Tuple[List[str], bool]:
                 else:
                     has_two_fields.append(False)
                 audio_key = cfg.get('audio_key', 'audio_filepath')
-                audio_file = Path(item[audio_key])
-                if not audio_file.is_file() and not audio_file.is_absolute():
```
Comment (titu1994) [inline nemo/collections/asr/parts/utils/transcribe_utils.py:231]: In reluctant to change any of this. It has severe negative interaction with tarred datasets and filename size limits.

File: tests/collections/asr/utils/test_audio_utils.py [hunk 4]
```diff
@@ -39,6 +41,150 @@
 )
 
 
+class TestAudioSegment:
+    @pytest.mark.unit
+    @pytest.mark.parametrize(
+        "num_channels, channel_selectors", [(1, [None, 'average', 0]), (3, [None, 'average', 0, 1, [0, 1]]),]
```
Comment (titu1994) [inline tests/collections/asr/utils/test_audio_utils.py:47]: Can you add a few tests - 
- File path longer than 255 chars 
- filename longer than 255 chars
- two audio files incorrectly passed as different channel (raise appropriate error)

**Review by titu1994 — CHANGES_REQUESTED** [review]
Code looks ok, but there are sections that are very sensitive to changes and break a lot of underlying data loaders so I'm reluctant to merge this with changes to critical sections

File: tests/collections/asr/utils/test_audio_utils.py (hunk 4 above)
Comment (titu1994) [reply tests/collections/asr/utils/test_audio_utils.py:47]: Ok that seems like a reasonable error for (3). Let's add (1-2) then merge if they pass

**Review by tango4j — CHANGES_REQUESTED** [review]

File: nemo/collections/asr/parts/preprocessing/segment.py [hunk 5]
```diff
@@ -199,51 +201,111 @@ def from_file(
...
+                    raise RuntimeError(
+                        f'Channel cannot be selected: channel_selector={channel_selector}, num_audio_files={len(audio_file)}'
+                    )
+                # Select only a single file
+                audio_file = [audio_file[channel_selector]]
+                # Reset the channel selector since we applied it here
+                channel_selector = None
+
+            # Load all files
+            for a_file in audio_file:
```
[the remark this answers is not in the document]
Comment (tango4j) [reply nemo/collections/asr/parts/preprocessing/segment.py]: Thanks after code-QL is addressed, we can merge I guess

**Review by tango4j — APPROVED** [review]
AudioSegment.from_file() works as expected with real-life multichannel files and 
all my requested changes are reflected.
+ Need to check other reviewers before getting merged

File: nemo/collections/asr/parts/utils/transcribe_utils.py [hunk 6]
```diff
@@ -227,10 +226,8 @@ def prepare_audio_data(cfg: DictConfig) -> Tuple[List[str], bool]:
...
-                    audio_file = manifest_dir / audio_file
-                filepaths.append(str(audio_file.absolute()))
+                audio_file = get_full_path(audio_file=item[audio_key], manifest_file=cfg.dataset_manifest)
```
Comment (fayejf) [inline nemo/collections/asr/parts/utils/transcribe_utils.py:229]: so we need both audio_dir and manifest to locate the path of the audio file just like in your example file? 

`{"audio_filepath": ["61-70968-3-4-5_ch_0.wav", "61-70968-3-4-5_ch_1.wav", "61-70968-3-4-5_ch_2.wav"], "text": "he was like unto my father in a way and yet was not my father", "offset": 0, "duration": 4.32}`

**Review by fayejf — CHANGES_REQUESTED** [review]
need some clarification before I approve. Thanks

File: nemo/collections/asr/parts/utils/transcribe_utils.py (hunk 6 above)
Comment (fayejf) [reply nemo/collections/asr/parts/utils/transcribe_utils.py:229]: oh I see! Thanks for clarification!

**Review by fayejf — APPROVED** [review]
LGTM! Thanks!

File: examples/asr/transcribe_speech.py (hunk 1 above)
Comment (titu1994) [reply examples/asr/transcribe_speech.py:109]: Indeed union works now with the specifics that only primitive types are in the union

File: tests/collections/asr/utils/test_audio_utils.py (hunk 4 above)
Comment (titu1994) [reply tests/collections/asr/utils/test_audio_utils.py:48]: Perfect thanks !

**Review by titu1994 — APPROVED** [review]
Changes look great, thanks !

**Pull request merged**